# Tasca M5 T02
Objectivos

- Preprocesamiento de los datos.
- Transformación y creación de Dummys y enumeraciones.
- División de los datos en Train y Test.
- Estudio de los datos utilizando herramientas anteriores (gráficas, hipótesis…).

Duración: 2 días

Descripción:
Ejercicios de Train-Test con ScikitLearn a partir de un dataset con información sobre las viviendas de Boston (Concerns housing values in suburbs of Boston).

## Ejercicio 1
- Parte el conjunto de datos adjunto en train y test. Estudia ambos conjuntos por separado, a nivel descriptivo.
- También adjunto encontrarás una descripción de las diferentes variables del dataset.

__1.1.- EN PRIMER LUGAR: PARTIR EL CONJUNTO EN TRAIN/TEST__ : 

Voy a separar el conjutno de datos en __TRAIN/TEST__, como me pide el ejercicio, pero a su vez: 
- En el mismo código voya a entrenar a un modelo para que me permita realizar predicciones. 
- También en el mismo código voy a calcular la precisón de esas predicciones.

In [151]:
#1.- CARGO LA TABLA CSV y la almaceno en una variable llamada HOUSTON
import pandas as pd

HOUSTON = pd.read_csv('231206 housing data.csv', header=None)

# Pido una visualización rápida de los datos para ver su aspecto con el método to_string()
print('VISUALIZACIÓN RÁPIDA DE LOS DATOS')
print(HOUSTON.to_string())
print()

# Pido que me de la forma de esta tabla original
print('La forma de la matriz original de datos es:', HOUSTON.shape)

VISUALIZACIÓN RÁPIDA DE LOS DATOS
           0      1      2   3       4      5      6        7   8      9     10      11     12    13
0     0.00632   18.0   2.31   0  0.5380  6.575   65.2   4.0900   1  296.0  15.3  396.90   4.98  24.0
1     0.02731    0.0   7.07   0  0.4690  6.421   78.9   4.9671   2  242.0  17.8  396.90   9.14  21.6
2     0.02729    0.0   7.07   0  0.4690  7.185   61.1   4.9671   2  242.0  17.8  392.83   4.03  34.7
3     0.03237    0.0   2.18   0  0.4580  6.998   45.8   6.0622   3  222.0  18.7  394.63   2.94  33.4
4     0.06905    0.0   2.18   0  0.4580  7.147   54.2   6.0622   3  222.0  18.7  396.90   5.33  36.2
5     0.02985    0.0   2.18   0  0.4580  6.430   58.7   6.0622   3  222.0  18.7  394.12   5.21  28.7
6     0.08829   12.5   7.87   0  0.5240  6.012   66.6   5.5605   5  311.0  15.2  395.60  12.43  22.9
7     0.14455   12.5   7.87   0  0.5240  6.172   96.1   5.9505   5  311.0  15.2  396.90  19.15  27.1
8     0.21124   12.5   7.87   0  0.5240  5.631  100.0   6

In [199]:
# 2.- CONVIERTO el DataFrame HOUSTON en una matriz NumPy llamada data, y compruebo el tipo de elmentos.
data = HOUSTON.values 
print('La tabla HOUSTON es del tipo:',type(HOUSTON))
print('La tabla data es del tipo:', type(data))

# 3.- IDENTIFICO EN LA MATRIZ "data" EN ELEMENTOS DE ENTRADA Y SALIDA
# Los datos de salida (y) son la última columna
X, y = data[:, :-1], data[:, -1]

# 4.- COMPRUEBO LA FORMA DE LOS DOS ELEMENTOS DE ENTRADA Y SALIDA OBTENIDOS
print('La forma de los elementos de entrada y salida es', X.shape, y.shape)

# 5 .- DIVIDO LOS DATOS PARA EL TRAINTTEST (split into train test sets) 
# Para ello, primero importo: train_test_split
from sklearn.model_selection import train_test_split

# Luego tomo la decisión: el 67 por ciento se utilizará para entrenar el modelo y el 33 por ciento para evaluarlo.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=1)

# 6.- COMPRUEBO LA FORMA DE AMBAS DIVISIONES 
print('La forma para ambas divisones es:', X_train.shape, X_test.shape, y_train.shape, y_test.shape)

# 7.- AJUSTO/ENTRENO EL MODELO (fit the model) CON UN MODELO DE REGRESIÓN ALEATORIA EN EL CONJUNTO DE ENTRENAMIENTO:
# Primero importo la clase RandomForestRegressor desde el módulo sklearn.ensemble. 
# RandomForestRegressor es un modelo de regresión basado en la técnica de Random Forest (árboles de decisión). 
# se utiliza específicamente para resolver problemas de regresión, donde la variable de destino es continua.
# (si la variable hubiese sido discreta o categórica hubiese ultilizado "RandomForestClassifier" en su lugar)
from sklearn.ensemble import RandomForestRegressor

# Controlo la aleatorización durante la división con random_state
# El parámetro random_state se establece en 1, ==> se fijará la semilla para la generación de números aleatorios. 
# Fijar la semilla garantiza que los resultados sean reproducibles; es decir: 
# que si ejecuto el mismo código varias veces con la misma semilla, debería obtener los mismos resultados 
model = RandomForestRegressor(random_state=1)
model.fit(X_train, y_train)
# RandomForestRegressor crea el objeto que representa el modelo, mientras .fit()entrena o ajusta el modelo y lo devuelve

# 8.- HAGO PREDICCIONES:
# Primero tengo que importar mean_absolute_error
from sklearn.metrics import mean_absolute_error

yhat = model.predict(X_test)
# yhat = model.predict(X_test): Este código utiliza el modelo (model) previamente entrenado 
# para hacer predicciones sobre un conjunto de datos de prueba (X_test). 
# El resultado de estas predicciones se almacena en la variable yhat.

# 9.- EVALUO PREDICCIONES: 
# Para hacer predicciones y evaluarlas utilizo el mean absolute error (MAE / error medio absoluto).

mae = mean_absolute_error(y_test, yhat)
# "mae = mean_absolute_error(y_test, yhat)" calcula el error absoluto medio (MAE) entre las predicciones (yhat) 
# y las etiquetas reales (y_test). Explicado de forma más detallada:
# mean_absolute_error: calcula el error absoluto medio entre dos conjuntos de valores. La fórmula para el MAE es 
#la suma de las diferencias absolutas entre las predicciones y los valores reales, dividida por el número total de observaciones.
# El resultado (mae) me da una medida del error promedio absoluto de las predicciones en relación con las verdaderas etiquetas. 
# Un MAE más bajo indica que las predicciones están más cerca de los valores reales, 
# mientras que un MAE más alto indica mayores discrepancias entre las predicciones y las observaciones reales.

print('El resultado MAE (error absoluto medio) es : %.3f' % mae)

# En resumen, estas líneas de código realizan predicciones utilizando un modelo entrenado (model) 
# en un conjunto de datos de prueba (X_test), calculan la precisión de las predicciones utilizando 
# la función "mean_absolute_error" y finalmente imprimen la precisión en la consola. 
# La precisión es una medida importante para evaluar el rendimiento de un modelo de clasificación.

La tabla HOUSTON es del tipo: <class 'pandas.core.frame.DataFrame'>
La tabla data es del tipo: <class 'numpy.ndarray'>
La forma de los elementos de entrada y salida es (506, 13) (506,)
La forma para ambas divisones es: (339, 13) (167, 13) (339,) (167,)
El resultado MAE (error absoluto medio) es : 2.171


__MIS CONCLUSIONES SOBRE EL TRAIN/TEST__

He divido los datos y he entrenado al modelo. La división de los datos estimo que es correcta porque he tomado un división 67/33, ampliamente aceptado. El entrenamiento de datos y las predicciones también tengo que pensar que los estoy haciendo bien, aunque no encuentro ningún sistema que me permita confirmar si lo que he hecho está correcto o no. Por último, al medir la precisión del modelo me da un valor de 2.171. No tengo ninguna referencia con la que comparar este valor; solo sé que cuanto más bajo sea el MAE más preciso es el sistema. ¿Cómo saber si 2.171 es un valor alto o bajo? Pues la verdad es que no tengo ni idea y tampoco he encontrado en el temario recomendado ninguna aclaración al respecto. 

__1.2.- EN SEGUNDO LUGAR: ESTUDIAR AMBOS CONJUNTOS POR SEPARADO A NIVEL DESCRIPTIVO__ :

In [153]:
# Primero les asigno los nombres de las columnas, para poder transformar los dos objetos Numpy en Dataframes:
column_names = [
    "CRIM", "ZN", "INDUS", "CHAS", "NOX", "RM", "AGE", "DIS", "RAD",
    "TAX", "PTRATIO", "B", "LSTAT", "MEDV"
]

# Luego convierot las matrices en DataFrames
X_train_df = pd.DataFrame(X_train, columns=column_names[:-1])
X_test_df = pd.DataFrame(X_test, columns=column_names[:-1])
y_train_df = pd.DataFrame(y_train, columns=["MEDV"])
y_test_df = pd.DataFrame(y_test, columns=["MEDV"])

In [154]:
# ANALIZO EL GRUPO DE ENTRENAMIENTO
print("Descripción estadística del GRUPO DE ENTRENAMIENTO:")
X_train_df.describe()

Descripción de X_train:


CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  339.000000  339.000000  339.000000  339.000000  339.000000  339.000000   
mean     3.793986   10.963127   11.377493    0.079646    0.556238    6.247490   
std      8.669966   23.058729    6.874160    0.271145    0.117996    0.669953   
min      0.006320    0.000000    0.460000    0.000000    0.392000    3.863000   
25%      0.085770    0.000000    5.640000    0.000000    0.448000    5.873000   
50%      0.253870    0.000000    9.900000    0.000000    0.538000    6.167000   
75%      3.812340   12.500000   18.100000    0.000000    0.627500    6.582500   
max     73.534100  100.000000   27.740000    1.000000    0.871000    8.398000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  339.000000  339.000000  339.000000  339.000000  339.000000  339.000000   
mean    69.173451    3.799550    9.622419  408.890855   18.418584  356.552537   
std     28.246657    2.114897    8.768838  169.654094    2.176714   92.095470   
min      6.000000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.750000    2.094450    4.000000  279.000000   17.000000  375.990000   
50%     79.700000    3.262800    5.000000  330.000000   19.000000  391.450000   
75%     94.300000    5.033750   24.000000  666.000000   20.200000  396.060000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT  
count  339.000000  
mean    13.087522  
std      7.340139  
min      1.730000  
25%      7.415000  
50%     12.030000  
75%     17.135000  
max     37.970000

In [155]:
print("Información del GRUPO DE ENTRENAMIENTO:")
X_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 339 entries, 0 to 338
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     339 non-null    float64
 1   ZN       339 non-null    float64
 2   INDUS    339 non-null    float64
 3   CHAS     339 non-null    float64
 4   NOX      339 non-null    float64
 5   RM       339 non-null    float64
 6   AGE      339 non-null    float64
 7   DIS      339 non-null    float64
 8   RAD      339 non-null    float64
 9   TAX      339 non-null    float64
 10  PTRATIO  339 non-null    float64
 11  B        339 non-null    float64
 12  LSTAT    339 non-null    float64
dtypes: float64(13)
memory usage: 34.6 KB


In [200]:
print("Descripción estadística del GRUPO DE PRUEBA:")
X_test_df.describe()

Descripción estadística del GRUPO DE PRUEBA:


CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  167.000000  167.000000  167.000000  167.000000  167.000000  167.000000   
mean     3.247195   12.176647   10.648144    0.047904    0.551563    6.360036   
std      8.474939   23.898258    6.826662    0.214206    0.111736    0.761180   
min      0.010960    0.000000    1.250000    0.000000    0.385000    3.561000   
25%      0.076945    0.000000    4.675000    0.000000    0.458000    5.979500   
50%      0.261690    0.000000    8.560000    0.000000    0.532000    6.310000   
75%      3.122525   20.000000   18.100000    0.000000    0.624000    6.630500   
max     88.976200   95.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  167.000000  167.000000  167.000000  167.000000  167.000000  167.000000   
mean    67.359880    3.785894    9.401198  406.910180   18.530539  356.920659   
std     27.994376    2.093240    8.605168  166.745525    2.145399   89.921726   
min      2.900000    1.178100    1.000000  188.000000   13.000000    3.650000   
25%     41.700000    2.101800    4.000000  281.000000   17.400000  373.715000   
50%     73.300000    3.092300    5.000000  337.000000   19.100000  390.680000   
75%     93.150000    5.307650   16.000000  666.000000   20.200000  396.660000   
max    100.000000   10.710300   24.000000  711.000000   22.000000  396.900000   

            LSTAT  
count  167.000000  
mean    11.771138  
std      6.653123  
min      1.920000  
25%      6.605000  
50%     10.240000  
75%     15.695000  
max     30.810000

In [201]:
print("Información del GRUPO DE PRUEBA:")
X_test_df.info()

Información del GRUPO DE PRUEBA:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 167 entries, 0 to 166
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     167 non-null    float64
 1   ZN       167 non-null    float64
 2   INDUS    167 non-null    float64
 3   CHAS     167 non-null    float64
 4   NOX      167 non-null    float64
 5   RM       167 non-null    float64
 6   AGE      167 non-null    float64
 7   DIS      167 non-null    float64
 8   RAD      167 non-null    float64
 9   TAX      167 non-null    float64
 10  PTRATIO  167 non-null    float64
 11  B        167 non-null    float64
 12  LSTAT    167 non-null    float64
dtypes: float64(13)
memory usage: 17.1 KB


In [158]:
print("Descripción estdiística de la variable de SALIDA en el GRUPO DE ENTRENAMIENTO:")
y_train_df.describe()


Descripción de y_train:


MEDV
count  339.000000
mean    22.289971
std      9.093571
min      5.000000
25%     16.900000
50%     20.600000
75%     25.000000
max     50.000000

In [203]:
print("Descripción estadiística de la variable de SALIDA en el GRUPO DE PRUEBA:")
y_test_df.describe()

Descripción estadiística de la variable de SALIDA en el GRUPO DE PRUEBA:


MEDV
count  167.000000
mean    23.025749
std      9.412079
min      6.300000
25%     17.500000
50%     22.100000
75%     25.200000
max     50.000000

__MIS CONCLUSIONES SOBRE EL ANÁLISIS ESTADÍSTICO EN LOS CONJUNTOS DE ENTRENO Y PRUEBA__

Encuentro mucha disparidad entre los análisisi estadísticos de cada una de las variables. Según la literatura consultada, unos datos tan heterogéneos no son recomendables para crear modelos de entreno. Intuyo que los datos van a necesitar algún tipo de preprocesado con el fin de mejorar su eficiencia en los modelos Train/test. Es lo que voy a intentar hacer en el segundo ejercicio.

__.........................................................................................................................................................................................................................................................................__

## Ejercicio 2
Aplica algún proceso de transformación (estandarizar los datos numéricos, crear columnas dummies, polinomios...).

### 2.1.- ESTANDARIZACIÓN:
- Este método va bien cuando estamos utilizando modelos sensibles a la escala de las características (como la regresión lineal, máquinas de soporte vectorial, k-means, etc.). 
- según ChantGPT Puede resultar beneficioso estandarizar los datos numéricos para que tengan una media de 0 y una desviación estándar de 1.

In [160]:
# Aplico el código de estandarización que aparece en el ScikitLearn Cheatsheet.pdf
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(X_train)
standardized_X_train = scaler.transform(X_train)standardized_X_test = scaler.transform(X_test)

In [209]:
import numpy as np
print("Conjunto de ENTRENAMIENTO estandarizado:\n")
print(np.round(standardized_X_train, 3))

print("\nConjunto de PRUEBA estandarizado:\n")
print(np.round(standardized_X_test, 3))

Conjunto de ENTRENAMIENTO estandarizado:

[[ 2.293 -0.476  0.979 ...  0.82   0.439  1.447]
 [-0.408 -0.476  1.532 ...  1.28   0.387  0.56 ]
 [-0.434  0.61  -0.95  ...  0.268  0.439 -0.869]
 ...
 [ 0.24  -0.476  0.979 ...  0.82   0.439  0.857]
 [-0.4   -0.476 -0.754 ... -0.469  0.22  -0.301]
 [-0.429 -0.476 -0.789 ...  0.36   0.439 -0.589]]

Conjunto de PRUEBA estandarizado:

[[-0.433  0.957 -1.34  ... -0.009  0.439 -0.758]
 [-0.435  1.913 -1.107 ... -0.377  0.439 -0.806]
 [-0.412 -0.476 -0.651 ... -0.239  0.394  0.779]
 ...
 [ 0.953 -0.476  0.979 ...  0.82  -0.707  0.138]
 [-0.436  2.998 -1.401 ... -0.101 -0.163 -0.021]
 [-0.436  3.216 -1.053 ... -0.239  0.39  -0.918]]


In [210]:
# Comparo el GRUPO DE ENTRENAMIENTO antes y después de la estandarización.
print("Primeras 5 filas del GRUPO DE ENTRENAMIENTO originales:\n")
print(np.round(X_train[:5, :],3))

print("\nPrimeras 5 filas del GRUPO DE ENTRENAMIENTO estandarizados:\n")
print(np.round(standardized_X_train[:5, :],3))

Primeras 5 filas del GRUPO DE ENTRENAMIENTO originales:

[[2.3648e+01 0.0000e+00 1.8100e+01 0.0000e+00 6.7100e-01 6.3800e+00
  9.6200e+01 1.3860e+00 2.4000e+01 6.6600e+02 2.0200e+01 3.9690e+02
  2.3690e+01]
 [2.5900e-01 0.0000e+00 2.1890e+01 0.0000e+00 6.2400e-01 5.6930e+00
  9.6000e+01 1.7880e+00 4.0000e+00 4.3700e+02 2.1200e+01 3.9211e+02
  1.7190e+01]
 [3.7000e-02 2.5000e+01 4.8600e+00 0.0000e+00 4.2600e-01 6.3020e+00
  3.2200e+01 5.4010e+00 4.0000e+00 2.8100e+02 1.9000e+01 3.9690e+02
  6.7200e+00]
 [5.8210e+00 0.0000e+00 1.8100e+01 0.0000e+00 7.1300e-01 6.5130e+00
  8.9900e+01 2.8020e+00 2.4000e+01 6.6600e+02 2.0200e+01 3.9382e+02
  1.0290e+01]
 [2.5200e-01 0.0000e+00 1.0590e+01 0.0000e+00 4.8900e-01 5.7830e+00
  7.2700e+01 4.3550e+00 4.0000e+00 2.7700e+02 1.8600e+01 3.8943e+02
  1.8060e+01]]

Primeras 5 filas del GRUPO DE ENTRENAMIENTO estandarizados:

[[ 2.293 -0.476  0.979 -0.294  0.974  0.198  0.958 -1.143  1.642  1.518
   0.82   0.439  1.447]
 [-0.408 -0.476  1.532 -0.294  0.5

In [163]:
# SOBRE LAS TABLAS ESTANDARIZADAS VUELVO A COMPROBAR SUS FORMAS, A HACER PREDICCIONES 
# Y A COMPROBAR LA EFICACIA DE LAS PREDICCIONES: 

# 6.- COMPRUEBO LA FORMA DE AMBAS DIVISIONES 
print('La forma para ambas divisones es:', standardized_X_train.shape, standardized_X_test.shape, y_train.shape, y_test.shape)

# 7.- AJUSTO/ENTRENO EL MODELO (fit the model) CON UN MODELO DE REGRESIÓN ALEATORIA EN EL CONJUNTO DE ENTRENAMIENTO:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(random_state=1)
model.fit(standardized_X_train, y_train)

# 8.- HAGO PREDICCIONES:
from sklearn.metrics import mean_absolute_error
yhat = model.predict(standardized_X_test)

# 9.- EVALUO PREDICCIONES: 
mae = mean_absolute_error(y_test, yhat)

print('El resultado MAE (error absoluto medio) es : %.3f' % mae)

# Observo que el resultado es el mismo que antes de estandarizar 
# ==> No alcanzo a entender el si la estandarización me ha beneficiado en algo.

La forma para ambas divisones es: (339, 13) (167, 13) (339,) (167,)
El resultado MAE (error absoluto medio) es : 2.171


__MIS CONCLUSIONES SOBRE EL PREPROCESADO DE ESTANDARIZACIÓN__

Sin ninguna duda las tablas una vez estandarizadas presentan mucho mejor aspecto que sin estandarizar, quiero decir, que los datos parecen más homogéneos y mucho más predispuestos a ser comparados entre si. 

No onstante al intentar hacer predicciones e intentar evaluar la eficiencia de esas predicciones, observo que el resultado MAE es exactamente el mismo. 

Por lo tanto conluyo que el proceso de estandarización, en este caso en concreto, no aporta ninguna ventaja. 

### 2.2.- CREAR COLUMNAS DUMMIES PARA COLUMNAS CATEGÓRICAS:


No alcanzo a entender el sentido de crear columnas Dummies para esta tabla. 
Según la literatura consultada, las columnas Dummies se utilizan para transformar los datos categóricos, 
con los que no se puede trabajar, en datos numéricos binarios (0,1) con los que sí se puede trabajar. 
En esta tabla no hay ninguna variable que contenga datos categóricos, únicamente la variable CHAS, 
que la leyenda indica que es "Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)"", 
y por lo tanto ya nos indica que es una columna Dummy.

No obstante, a pesar de encontrarlo absurdo, voy a intentar aplicar los códigos que recomienda la literatura 
para ver cómo se comporta e intentar llegar a alguna conclusión.





__EN PRIMER LUGAR__: analizo el código que recomienda el ScikitLearn Cheatsheet.pdf para ENCODING CATEGORICAL FEATURES: 
- from sklearn.preprocessing import LabelEncoder
- enc = LabelEncoder()
- y = enc.fit_transform(y)

Encuentro que es un código que utiliza LabelEncoder de scikit-learn para transformar las etiquetas categóricas 
en números enteros. 

El objetivo principal de LabelEncoder es convertir etiquetas de texto en números para que puedan ser utilizadas como entradas en modelos de aprendizaje automático.

Y eso es precisamente lo que yo ya tengo en la columna CHAS, por lo que no le encuentro sentido a aplicarlo. 

Además, en este caso el Cheatsheet parece aplicar el código sobre la variable y, que representa las etiquetas de salida. Por lo que aún le encuentro menos sentido.


__EN SEGUNDO LUGAR__: consulto el sitio oficial de scikit-learn:https://scikit-learn.org/stable/modules/preprocessing.html

En concreo el apartado __6.3.4. Codificación de características categóricas__. 

Me encuentro en el mismo caso. Todos los ejemplos que ofrencen parten de variables categóricas que son "Nombres" para ser transformadas en datos binarios (0,1), por lo que no encuentro aplicable ninguno de los códigos que me recomienda.


__FINALMENTE__: hago la consulta a mi amigo CHATGPT, que me recomienda el siguiente código

In [172]:
# CREAR COLUMNAS DUMMIES
X_train_df_dummy = pd.concat([X_train_df, pd.get_dummies(X_train_df['CHAS'], prefix='CHAS', drop_first=True)], axis=1)
X_test_df_dummy = pd.concat([X_test_df, pd.get_dummies(X_test_df['CHAS'], prefix='CHAS', drop_first=True)], axis=1)

# pd.get_dummies: función de pandas que crea columnas dummies a partir de la columna categórica'CHAS'. 
# (Crea una columna dummy para cada categoría única en 'CHAS').
# prefix='CHAS': Agrega el prefijo 'CHAS' a todas las nuevas columnas dummies para identificar su origen.
# drop_first=True: Se utiliza para evitar la multicolinealidad al eliminar la primera columna dummy ==> 
# esto significa que si hay dos categorías en 'CHAS', solo necesito una columna dummy para representarlas, 
# ya que la otra se puede inferir.
# pd.concat: Combina estas nuevas columnas dummies con el DataFrame original X_train_df y X_test_df, respectivamente.

# Elimino la columna original CHAS de las dos tablas, ya que ahora está representada por las dummies creadas.
X_train_df_dummy.drop('CHAS', axis=1, inplace=True)
X_test_df_dummy.drop('CHAS', axis=1, inplace=True)

In [214]:
print('\nLa tabla de ETRENAMIENTO CON LAS NUEVAS COLUMNAS DUMMIES es: ')
X_train_df_dummy


La tabla de ETRENAMIENTO CON LAS NUEVAS COLUMNAS DUMMIES es: 


CRIM    ZN  INDUS    NOX     RM   AGE     DIS   RAD    TAX  PTRATIO  \
0    23.64820   0.0  18.10  0.671  6.380  96.2  1.3861  24.0  666.0     20.2   
1     0.25915   0.0  21.89  0.624  5.693  96.0  1.7883   4.0  437.0     21.2   
2     0.03659  25.0   4.86  0.426  6.302  32.2  5.4007   4.0  281.0     19.0   
3     5.82115   0.0  18.10  0.713  6.513  89.9  2.8016  24.0  666.0     20.2   
4     0.25199   0.0  10.59  0.489  5.783  72.7  4.3549   4.0  277.0     18.6   
..        ...   ...    ...    ...    ...   ...     ...   ...    ...      ...   
334   0.03548  80.0   3.64  0.392  5.876  19.1  9.2203   1.0  315.0     16.4   
335   0.09164   0.0  10.81  0.413  6.065   7.8  5.2873   4.0  305.0     19.2   
336   5.87205   0.0  18.10  0.693  6.405  96.0  1.6768  24.0  666.0     20.2   
337   0.33045   0.0   6.20  0.507  6.086  61.5  3.6519   8.0  307.0     17.4   
338   0.08014   0.0   5.96  0.499  5.850  41.5  3.9342   5.0  279.0     19.2   

          B  LSTAT  CHAS_1.0  
0    396.90  23.69         0  
1    392.11  17.19         0  
2    396.90   6.72         0  
3    393.82  10.29         0  
4    389.43  18.06         0  
..      ...    ...       ...  
334  395.18   9.25         0  
335  390.91   5.52         0  
336  396.90  19.37         0  
337  376.75  10.88         0  
338  396.90   8.77         0  

[339 rows x 13 columns]

In [213]:
print('\nLa tabla de PRUEBA CON LAS NUEVAS COLUMNAS DUMMIES es: ')
X_test_df_dummy


La tabla de PRUEBA CON LAS NUEVAS COLUMNAS DUMMIES es: 


CRIM    ZN  INDUS    NOX     RM   AGE     DIS   RAD    TAX  PTRATIO  \
0     0.04932  33.0   2.18  0.472  6.849  70.3  3.1827   7.0  222.0     18.4   
1     0.02543  55.0   3.78  0.484  6.696  56.4  5.7321   5.0  370.0     17.6   
2     0.22927   0.0   6.91  0.448  6.030  85.5  5.6894   3.0  233.0     17.9   
3     0.05789  12.5   6.07  0.409  5.878  21.4  6.4980   4.0  345.0     18.9   
4     3.67822   0.0  18.10  0.770  5.362  96.2  2.1036  24.0  666.0     20.2   
..        ...   ...    ...    ...    ...   ...     ...   ...    ...      ...   
162   0.19073  22.0   5.86  0.431  6.718  17.5  7.8265   7.0  330.0     19.1   
163   7.02259   0.0  18.10  0.718  6.006  95.3  1.8746  24.0  666.0     20.2   
164  12.04820   0.0  18.10  0.614  5.648  87.6  1.9512  24.0  666.0     20.2   
165   0.01965  80.0   1.76  0.385  6.230  31.5  9.0892   1.0  241.0     18.2   
166   0.01870  85.0   4.15  0.429  6.516  27.7  8.5353   4.0  351.0     17.9   

          B  LSTAT  CHAS_1.0  
0    396.90   7.53         0  
1    396.90   7.18         0  
2    392.74  18.80         0  
3    396.21   8.10         0  
4    380.79  10.19         0  
..      ...    ...       ...  
162  393.74   6.56         0  
163  319.98  15.70         0  
164  291.55  14.10         0  
165  341.60  12.93         0  
166  392.43   6.36         0  

[167 rows x 13 columns]

__MIS CONCLUSIONES SOBRE LA CREACIÓN DE COLUMNAS DUMMIES__

El resultado lo encuentro estúpido porque me da un columna (0,1) en una tabla en la que ya tenía una columna (0,1), por lo tanto, no encuentro el sentido de realizar este preprocesado.

### 2.3.- GENERAR ENTIDADES POLINÓMICAS:

Según la literatura consultada, resulta útil agregar complejidad a un modelo considerando características no lineales de los datos de entrada (para empezar, no entiendo el argumento).

__PRUEBA ÚNICA__: Intento en primer lugar estudiar el código que me ofrece el ScikitLearn Cheatsheet.pdf en el apartado: Generating Polynomial Features

In [182]:
from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(5)
poly.fit_transform(X_train_df)

array([[1.00000000e+00, 2.36482000e+01, 0.00000000e+00, ...,
        2.09438915e+09, 1.25009017e+08, 7.46148557e+06],
       [1.00000000e+00, 2.59150000e-01, 0.00000000e+00, ...,
        7.80986392e+08, 3.42382395e+07, 1.50099548e+06],
       [1.00000000e+00, 3.65900000e-02, 2.50000000e+01, ...,
        4.78046361e+07, 8.09390665e+05, 1.37039689e+04],
       ...,
       [1.00000000e+00, 5.87205000e+00, 0.00000000e+00, ...,
        1.14485652e+09, 5.58726901e+07, 2.72676747e+06],
       [1.00000000e+00, 3.30450000e-01, 0.00000000e+00, ...,
        1.82807163e+08, 5.27920884e+06, 1.52455984e+05],
       [1.00000000e+00, 8.01400000e-02, 0.00000000e+00, ...,
        1.06257839e+08, 2.34789933e+06, 5.18797610e+04]])

In [180]:
poly.fit_transform(X_test_df)

array([[1.00000000e+00, 4.93200000e-02, 3.30000000e+01, ...,
        6.72584921e+07, 1.27603035e+06, 2.42088902e+04],
       [1.00000000e+00, 2.54300000e-02, 5.50000000e+01, ...,
        5.83089916e+07, 1.05482126e+06, 1.90819266e+04],
       [1.00000000e+00, 2.29270000e-01, 0.00000000e+00, ...,
        1.02490549e+09, 4.90610154e+07, 2.34849287e+06],
       ...,
       [1.00000000e+00, 1.20482000e+01, 0.00000000e+00, ...,
        2.38277717e+08, 1.15236351e+07, 5.57308367e+05],
       [1.00000000e+00, 1.96500000e-02, 8.00000000e+01, ...,
        2.52250072e+08, 9.54799013e+06, 3.61403725e+05],
       [1.00000000e+00, 1.87000000e-02, 8.50000000e+01, ...,
        3.96182919e+07, 6.42082248e+05, 1.04060421e+04]])

__MIS CONCLUSIONES SOBRE LA GENERACIÓN DE ENTIDADES POLINÓMICAS__:

Encuentro que me devuelve dos matrices totalmente idénticas tanto para el grupo de valores de entrenamiento como para el de prueba. 

Por lo tanto deduzco que este método no tiene sentido aplicarlo a nuestro grupo de datos.

__.........................................................................................................................................................................................................................................................................__

## Ejercicio 3
Resume las nuevas columnas generadas de forma estadística y gráfica

In [189]:
# Convertir matrices en DataFrames
standardized_X_train_df = pd.DataFrame(standardized_X_train, columns=column_names[:-1])
standardized_X_test_df = pd.DataFrame(standardized_X_test, columns=column_names[:-1])

### 3.1.- DATOS OBTENIDOS TRAS EL PROCESO DE ESTANDARIZACIÓN
Pido que me devuelva una vista de la tabla y una descripción estadística tanto del grupo de ENTRENO como del de PRUEBA.

En las vistas de ambas tablas encuentro que los datos están bastante armonizados, por lo que intuyo que la estandarización puede aportar beneficios en el momento de entrenar modelos.

Las descripciones estadísticas de las tablas me aportan datos y lecturas que no soy capaz de interpretar.

En cuanto a la representación gráfica, no tengo ni idea de que tipo de representación debería aplicar.

In [190]:
standardized_X_train_df

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
0    2.293384 -0.476146  0.979384 -0.294174  0.974028  0.198083  0.958220   
1   -0.408313 -0.476146  1.531539 -0.294174  0.575122 -0.828878  0.951129   
2   -0.434021  0.609644 -0.949516 -0.294174 -1.105374  0.081485 -1.310885   
3    0.234160 -0.476146  0.979384 -0.294174  1.330497  0.396898  0.734855   
4   -0.409140 -0.476146 -0.114728 -0.294174 -0.570671 -0.694342  0.125033   
..        ...       ...       ...       ...       ...       ...       ...   
334 -0.434150  2.998384 -1.127255 -0.294174 -1.393945 -0.555321 -1.775342   
335 -0.427662 -0.476146 -0.082677 -0.294174 -1.215710 -0.272794 -2.175980   
336  0.240040 -0.476146  0.979384 -0.294174  1.160750  0.235454  0.951129   
337 -0.400077 -0.476146 -0.754295 -0.294174 -0.417898 -0.241403 -0.272060   
338 -0.428991 -0.476146 -0.789260 -0.294174 -0.485797 -0.594187 -0.981155   

          DIS       RAD       TAX   PTRATIO         B     LSTAT  
0   -1.142854  1.642046  1.517731  0.819606  0.438752  1.446587  
1   -0.952398 -0.642130  0.165930  1.279694  0.386664  0.559736  
2    0.758201 -0.642130 -0.754947  0.267502  0.438752 -0.868776  
3   -0.472564  1.642046  1.517731  0.819606  0.405259 -0.381690  
4    0.262978 -0.642130 -0.778560  0.083467  0.357521  0.678438  
..        ...       ...       ...       ...       ...       ...  
334  2.566917 -0.984756 -0.554243 -0.928724  0.420048 -0.523586  
335  0.704502 -0.642130 -0.613274  0.359519  0.373615 -1.032502  
336 -1.005197  1.642046  1.517731  0.819606  0.438752  0.857172  
337 -0.069917 -0.185294 -0.601468 -0.468637  0.219634 -0.301191  
338  0.063762 -0.527921 -0.766753  0.359519  0.438752 -0.589077  

[339 rows x 13 columns]

In [191]:
standardized_X_train_df.describe()

CRIM            ZN         INDUS          CHAS           NOX  \
count  3.390000e+02  3.390000e+02  3.390000e+02  3.390000e+02  3.390000e+02   
mean   2.095996e-17 -1.047998e-17  3.746593e-15 -4.715992e-17 -2.984175e-15   
std    1.001478e+00  1.001478e+00  1.001478e+00  1.001478e+00  1.001478e+00   
min   -4.375179e-01 -4.761465e-01 -1.590541e+00 -2.941742e-01 -1.393945e+00   
25%   -4.283405e-01 -4.761465e-01 -8.358802e-01 -2.941742e-01 -9.186526e-01   
50%   -4.089231e-01 -4.761465e-01 -2.152520e-01 -2.941742e-01 -1.547904e-01   
75%    2.120063e-03  6.674891e-02  9.793844e-01 -2.941742e-01  6.048282e-01   
max    8.055764e+00  3.867017e+00  2.383811e+00  3.399346e+00  2.671500e+00   

                 RM           AGE           DIS           RAD           TAX  \
count  3.390000e+02  3.390000e+02  3.390000e+02  3.390000e+02  3.390000e+02   
mean  -1.281178e-15  1.630947e-15  4.401592e-16 -2.881995e-17  5.239991e-17   
std    1.001478e+00  1.001478e+00  1.001478e+00  1.001478e+00  1.001478e+00   
min   -3.564450e+00 -2.239799e+00 -1.264315e+00 -9.847559e-01 -1.309835e+00   
25%   -5.598052e-01 -8.304727e-01 -8.074249e-01 -6.421296e-01 -7.667534e-01   
50%   -1.203198e-01  3.732162e-01 -2.541699e-01 -5.279208e-01 -4.656974e-01   
75%    5.007896e-01  8.908555e-01  5.844375e-01  1.642046e+00  1.517731e+00   
max    3.214686e+00  1.092948e+00  3.943105e+00  1.642046e+00  1.783368e+00   

            PTRATIO             B         LSTAT  
count  3.390000e+02  3.390000e+02  3.390000e+02  
mean   2.312408e-14  4.433032e-15 -7.440787e-16  
std    1.001478e+00  1.001478e+00  1.001478e+00  
min   -2.677055e+00 -3.873797e+00 -1.549604e+00  
25%   -6.526722e-01  2.113697e-01 -7.739509e-01  
50%    2.675020e-01  3.794872e-01 -1.442868e-01  
75%    8.196065e-01  4.296179e-01  5.522322e-01  
max    1.647763e+00  4.387524e-01  3.394930e+00

In [193]:
standardized_X_test_df

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
0   -0.432551  0.957097 -1.339958 -0.294174 -0.714956  0.899167  0.039942   
1   -0.435310  1.912593 -1.106859 -0.294174 -0.613108  0.670455 -0.452880   
2   -0.411765 -0.476146 -0.650857 -0.294174 -0.918653 -0.325114  0.578854   
3   -0.431561  0.066749 -0.773235 -0.294174 -1.249660 -0.552331 -1.693796   
4   -0.013372 -0.476146  0.979384 -0.294174  1.814277 -1.323673  0.958220   
..        ...       ...       ...       ...       ...       ...       ...   
162 -0.416216  0.479349 -0.803829 -0.294174 -1.062938  0.703341 -1.832069   
163  0.372940 -0.476146  0.979384 -0.294174  1.372934 -0.360990  0.926310   
164  0.953454 -0.476146  0.979384 -0.294174  0.490249 -0.896146  0.653309   
165 -0.435978  2.998384 -1.401147 -0.294174 -1.453356 -0.026144 -1.335703   
166 -0.436088  3.215542 -1.052954 -0.294174 -1.079912  0.401382 -1.470431   

          DIS       RAD       TAX   PTRATIO         B     LSTAT  
0   -0.292100 -0.299503 -1.103228 -0.008550  0.438752 -0.758260  
1    0.915131 -0.527921 -0.229575 -0.376620  0.438752 -0.806014  
2    0.894911 -0.756338 -1.038294 -0.238594  0.393515  0.779402  
3    1.277812 -0.642130 -0.377152  0.221493  0.431249 -0.680490  
4   -0.803092  1.642046  1.517731  0.819606  0.263567 -0.395334  
..        ...       ...       ...       ...       ...       ...  
162  1.906903 -0.299503 -0.465697  0.313511  0.404389 -0.890606  
163 -0.911532  1.642046  1.517731  0.819606 -0.397702  0.356443  
164 -0.875259  1.642046  1.517731  0.819606 -0.706860  0.138141  
165  2.504836 -0.984756 -0.991070 -0.100568 -0.162599 -0.021492  
166  2.242545 -0.642130 -0.341733 -0.238594  0.390144 -0.917894  

[167 rows x 13 columns]

In [194]:
standardized_X_test_df.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  167.000000  167.000000  167.000000  167.000000  167.000000  167.000000   
mean    -0.063161    0.052705   -0.106257   -0.117239   -0.039672    0.168240   
std      0.978950    1.037940    0.994558    0.791174    0.948340    1.137849   
min     -0.436982   -0.476146   -1.475448   -0.294174   -1.453356   -4.015894   
25%     -0.429360   -0.476146   -0.976468   -0.294174   -0.833779   -0.400604   
50%     -0.408020   -0.476146   -0.410473   -0.294174   -0.205715    0.093443   
75%     -0.077561    0.392486    0.979384   -0.294174    0.575122    0.572542   
max      9.839500    3.649858    2.383811    3.399346    2.671500    3.785718   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  167.000000  167.000000  167.000000  167.000000  167.000000  167.000000   
mean    -0.064300   -0.006466   -0.025265   -0.011692    0.051509    0.004003   
std      0.992534    0.991223    0.982786    0.984309    0.987070    0.977840   
min     -2.349709   -1.241349   -0.984756   -1.303932   -2.493021   -3.837585   
25%     -0.974064   -0.803944   -0.642130   -0.754947   -0.468637    0.186631   
50%      0.146306   -0.334908   -0.527921   -0.424376    0.313511    0.371114   
75%      0.850083    0.714139    0.728376    1.517731    0.819606    0.436143   
max      1.092948    3.272484    1.642046    1.783368    1.647763    0.438752   

            LSTAT  
count  167.000000  
mean    -0.179606  
std      0.907743  
min     -1.523681  
25%     -0.884466  
50%     -0.388512  
75%      0.355761  
max      2.418030

### 3.1.- DATOS OBTENIDOS TRAS EL PROCESO OBTENER COLUMNAS DUMMIES
Pido que me devuelva una descripción estadística tanto del grupo de ENTRENO como del de PRUEBA depues de haber creado la columna dummie.

Veo que los resúmenes estadísticos son exactamente iguales que antes de haber creado esa columna, por lo que sigo sin entender el sentido de haber aplicado esta técnica de preprocesamiento.

En cuanto a la representación gráfica, al igual que en el caso anterior, no encuentro lógica a plantear una representación gráfica de estos resultados, por la sencilla de razon de que los resultados son absurdos.

In [195]:
X_train_df_dummy.describe()

CRIM          ZN       INDUS         NOX          RM         AGE  \
count  339.000000  339.000000  339.000000  339.000000  339.000000  339.000000   
mean     3.793986   10.963127   11.377493    0.556238    6.247490   69.173451   
std      8.669966   23.058729    6.874160    0.117996    0.669953   28.246657   
min      0.006320    0.000000    0.460000    0.392000    3.863000    6.000000   
25%      0.085770    0.000000    5.640000    0.448000    5.873000   45.750000   
50%      0.253870    0.000000    9.900000    0.538000    6.167000   79.700000   
75%      3.812340   12.500000   18.100000    0.627500    6.582500   94.300000   
max     73.534100  100.000000   27.740000    0.871000    8.398000  100.000000   

              DIS         RAD         TAX     PTRATIO           B       LSTAT  \
count  339.000000  339.000000  339.000000  339.000000  339.000000  339.000000   
mean     3.799550    9.622419  408.890855   18.418584  356.552537   13.087522   
std      2.114897    8.768838  169.654094    2.176714   92.095470    7.340139   
min      1.129600    1.000000  187.000000   12.600000    0.320000    1.730000   
25%      2.094450    4.000000  279.000000   17.000000  375.990000    7.415000   
50%      3.262800    5.000000  330.000000   19.000000  391.450000   12.030000   
75%      5.033750   24.000000  666.000000   20.200000  396.060000   17.135000   
max     12.126500   24.000000  711.000000   22.000000  396.900000   37.970000   

         CHAS_1.0  
count  339.000000  
mean     0.079646  
std      0.271145  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      0.000000  
max      1.000000

In [197]:
X_test_df_dummy.describe()

CRIM          ZN       INDUS         NOX          RM         AGE  \
count  167.000000  167.000000  167.000000  167.000000  167.000000  167.000000   
mean     3.247195   12.176647   10.648144    0.551563    6.360036   67.359880   
std      8.474939   23.898258    6.826662    0.111736    0.761180   27.994376   
min      0.010960    0.000000    1.250000    0.385000    3.561000    2.900000   
25%      0.076945    0.000000    4.675000    0.458000    5.979500   41.700000   
50%      0.261690    0.000000    8.560000    0.532000    6.310000   73.300000   
75%      3.122525   20.000000   18.100000    0.624000    6.630500   93.150000   
max     88.976200   95.000000   27.740000    0.871000    8.780000  100.000000   

              DIS         RAD         TAX     PTRATIO           B       LSTAT  \
count  167.000000  167.000000  167.000000  167.000000  167.000000  167.000000   
mean     3.785894    9.401198  406.910180   18.530539  356.920659   11.771138   
std      2.093240    8.605168  166.745525    2.145399   89.921726    6.653123   
min      1.178100    1.000000  188.000000   13.000000    3.650000    1.920000   
25%      2.101800    4.000000  281.000000   17.400000  373.715000    6.605000   
50%      3.092300    5.000000  337.000000   19.100000  390.680000   10.240000   
75%      5.307650   16.000000  666.000000   20.200000  396.660000   15.695000   
max     10.710300   24.000000  711.000000   22.000000  396.900000   30.810000   

         CHAS_1.0  
count  167.000000  
mean     0.047904  
std      0.214206  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      0.000000  
max      1.000000

__.........................................................................................................................................................................................................................................................................__

# MIS CONCLUSIONES SOBRE LA TASCA M5 T02

Estimada Lucía,

En esta segunda parte del Sprint 5 creo que he entendido bien lo que es un modelo de MACHINE LEARNING, cómo crear grupos de ENTRENAMIENTO Y PRUEBA, cómo ENTRENAR AL MODELO, como HACER PREDICCIONES  y cómo EVALUAR LAS PREDICCIONES.

Sin embargo tengo la sensación de que en la parte de __PREPROCESAMIENTO__ ha habido algo en lo que me he perdido. Por lo menos en cuanto al caso de estudio de los barrios de Boston que plantea el ejercicio. 

No he entendido qué se pretende con el preprocesamiento, ni cómo debe hacerse, ni qué ventajas aporta. Por mucho que vuelvo sobre los contenidos recomendados, no soy capaz de descodificar el sentido, ni el contenido, ni el objetivo. 

Entiendo que el preprocesamiento de datos no es un tema baladí, y que me voy a tener que enfrentar a ello millones de veces en mi desempeño como Data Science. Por ello te solicito una mentoría dedicada con el objeto de facilitarme el llegar a entender y descodificar el porqué y el cómo de estos procesos de preprocesado. 

Millones de gracias.

Esteban